app/controllers/api/v1/bulk_download_controller.rb

```diff
@@ -0,0 +1,308 @@
+          render json: e.message, status: 403 and return
+        end
+
+        # if this is a single-study download, allow for DirectoryListing downloads
+        if valid_accessions.size == 1 && params[:directory].present?
+          study_accession = valid_accessions.first
+          directories = self.class.find_matching_directories(params[:directory], study_accession)
+          directory_files = ::BulkDownloadService.get_requested_directory_files(directories)
+        else
+          directories = []
+          directory_files = []
+        end
+
+        files_requested = nil
+        # get requested files
+        # reference BulkDownloadService as ::BulkDownloadService to avoid NameError when resolving reference
+        if params[:file_ids]
+          files_requested = StudyFile.where(:id.in => file_ids)
+        else
+          files_requested = ::BulkDownloadService.get_requested_files(file_types: sanitized_file_types,
+                                                                      study_accessions: valid_accessions)
+        end
+
+        # determine quota impact & update user's download quota
+        # will throw a RuntimeError if the download exceeds the user's daily quota
+        begin
+          ::BulkDownloadService.update_user_download_quota(user: current_api_user, files: files_requested, directories: directories)
+        rescue RuntimeError => e
+          render json: {error: e.message}, status: 403 and return
+        end
+
+        # create maps to avoid Mongo timeouts when generating curl commands in parallel processes
+        bucket_map = ::BulkDownloadService.generate_study_bucket_map(valid_accessions)
+        pathname_map = ::BulkDownloadService.generate_output_path_map(files_requested, directories)
+
+        # generate curl config file
+        logger.info "Beginning creation of curl configuration for user_id, auth token: #{current_api_user.id}"
+        start_time = Time.zone.now
+        @configuration = ::BulkDownloadService.generate_curl_configuration(study_files: files_requested,
+                                                                           directory_files: directory_files,
+                                                                           user: current_api_user,
+                                                                           study_bucket_map: bucket_map,
+                                                                           output_pathname_map: pathname_map)
+        end_time = Time.zone.now
+        runtime = TimeDifference.between(start_time, end_time).humanize
+        logger.info "Curl configs generated for studies #{valid_accessions}, #{files_requested.size + directory_files.size} total files"
+        logger.info "Total time in generating curl configuration: #{runtime}"
+        send_data @configuration, type: 'text/plain', filename: 'cfg.txt'
+      end
+
+      private
+
+      def check_accession_permissions(valid_accessions)
```

Comment:
I think it would be better/more consistent if this were a class method, and we passed in any other parameters we need (like `current_api_user`).  This would make it more consistent with the structure of all the other private methods in this controller.

Reply:
concur